crates/primitives/starknet/src/execution/felt252_wrapper.rs

```diff
@@ -0,0 +1,404 @@
+//! # Felt252 - FieldElement wrapper.
+//!
+//! Starknet base type is a [`FieldElement`] from starknet-ff crate.
+//! Substrate primitives are passed back and forth between client
+//! and runtime using SCALE encoding: https://docs.substrate.io/reference/scale-codec/.
+//!
+//! The [`Felt252Wrapper`] implements the traits for SCALE encoding, and wrap
+//! the [`FieldElement`] type from starknet-ff.
+
+use alloc::string::String;
+
+use cairo_vm::felt::Felt252;
+use scale_codec::{Decode, Encode, EncodeLike, Error, Input, MaxEncodedLen, Output};
+use scale_info::build::Fields;
+use scale_info::{Path, Type, TypeInfo};
+use sp_core::{H256, U256};
+use starknet_api::hash::StarkFelt;
+use starknet_ff::{FieldElement, FromByteSliceError, FromStrError};
+
+///
+#[derive(Clone, Debug, PartialEq, Eq, Copy)]
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+pub struct Felt252Wrapper(pub FieldElement);
+
+impl Felt252Wrapper {
+    /// Initializes from a hex string.
+    ///
+    /// # Arguments
+    ///
+    /// * `value` - A valid hex string prefixed with '0x`, with or without padding zeros.
+    ///
+    /// # Errors
+    ///
+    /// Hex string may contain a value that overflows felt252.
+    /// If there if an overflow or invalid hex string,
+    /// returns [`Felt252WrapperError`].
+    pub fn from_hex_be(value: &str) -> Result<Self, Felt252WrapperError> {
+        let ff = FieldElement::from_hex_be(value)?;
+        Ok(Self(ff))
+    }
+
+    /// Initializes from a decimal string.
+    ///
+    /// # Arguments
+    ///
+    /// * `value` - A valid decimal string.
+    ///
+    /// # Errors
+    ///
+    /// Decimal string may contain a value that overflows felt252.
+    /// If there if an overflow or invalid character in the string,
+    /// returns [`Felt252WrapperError`].
+    pub fn from_dec_str(value: &str) -> Result<Self, Felt252WrapperError> {
+        let ff = FieldElement::from_dec_str(value)?;
+        Ok(Self(ff))
+    }
+
+    /// Inits from zero constant value.
+    pub fn zero() -> Felt252Wrapper {
+        Self(FieldElement::ZERO)
+    }
+
+    /// Inits from one constant value.
+    pub fn one() -> Felt252Wrapper {
+        Self(FieldElement::ONE)
+    }
+
+    /// Inits from two constant value.
+    pub fn two() -> Felt252Wrapper {
+        Self(FieldElement::TWO)
+    }
+
+    /// Inits from three constant value.
+    pub fn three() -> Felt252Wrapper {
+        Self(FieldElement::THREE)
+    }
+}
+
+impl Default for Felt252Wrapper {
+    fn default() -> Self {
+        Self(FieldElement::ZERO)
+    }
+}
+
+/// Array of bytes from [`Felt252Wrapper`].
+impl From<Felt252Wrapper> for [u8; 32] {
+    fn from(felt: Felt252Wrapper) -> Self {
+        felt.0.to_bytes_be()
+    }
+}
+
+/// [`Felt252Wrapper`] from bytes.
+/// Overflow may occur and return [`Felt252WrapperError::OutOfRange`].
+impl TryFrom<&[u8; 32]> for Felt252Wrapper {
```

Comment:
I would have expected `TryFrom<[u8; 32]>` to be implemented, rather than. `TryFrom<&[u8; 32]>`.

Reply:
Maybe both can exist